### DISCUSSION TITLE
Prevent encoding of #text values when XML is built

### DISCUSSION THREAD

**Discussion by gavinharriss** [NaturalIntelligence/fast-xml-parser#685] [Q&A] [answered]
Is there any way to turn off the encoding of `#text` elements when XML is being built from JSON?

I have a scenario where the destination system will take the raw value without unencoding text, so the encoded text shows up in the UI still encoded. This is beyond my control.

I'm wanting to capture a JSON representation of the raw source data for audit purposes. Though any `#text` field I use elsewhere is at risk of the same issue with user inputted values.

From this:

```
{
    "Note": {
        "Description": {
            "#text": "Raw source data"
        },
        "NoteText": {
            "#text": JSON.stringify(obj)
        }
    }
  }
```

The `XMLBuilder` outputs:

```
<Note>
	<Description>Raw source data</Description>
	<NoteText>{
  &quot;SomeProperty&quot;: {
    &quot;ChildProp&quot;: {
      &quot;x&quot;: 123,
      &quot;y&quot;: &quot;qwerty&quot;,
      &quot;z&quot;: &quot;xxx&quot;
    }
  }
}</NoteText>
</Note>
```

But what I need is:
```
<Note>
	<Description>Raw source data</Description>
	<NoteText>{
  "SomeProperty": {
    "ChildProp": {
      "x": 123,
      "y": "qwerty",
      "z": "xxx"
    }
  }
}</NoteText>
</Note>
```

Is there a way to achieve this?

I've tried experimenting with the `<![CDATA[]]>` tag, but unfortunately the destinations system doesn't care much for the `CDATA` tag.

**Comment by amitguptagwl** [comment]
What I understand here is that builder is not encoding the data. Can you please hardcode the value instead of `JSON.stringify` and debug?

**Reply by gavinharriss** [reply]
Thanks for taking a look, much appreciated :)

I've just tried passing `const chars = '"&<>'` in instead and the resulting output is `<NoteText>&quot;&amp;&lt;&gt;</NoteText>`, so it's definitely encoding the `#text` value.

**Comment by gavinharriss** [comment]
I did stumble across the `XmlBuilderOptions` `tagValueProcessor` option, but experimenting with this it only deals with the pre-unencoded value.

**Comment by gavinharriss** [comment]
I believe I've found what I need = setting the `processEntities: false` builder option. And if I need fine grained control around what chars are encoded it looks like I can use the `entities` setting.

**Reply by amitguptagwl** [reply]
This will help you to solve your problem.

https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/v4/5.Entities.md

**Reply by gavinharriss** [reply, marked as answer]
Unfortunately `entities` isn't exposed via the typescript definitions, but this is what worked for me in case anyone else has the same issue:

```
this.builder = new XMLBuilder({
    ignoreAttributes: false,
    // CargoWise only likes < and > chars encoded in text, so override encoding and provide our own encoder
    // NOTE: There appears to be an `entities` setting in js where we can set explicit chars to be encoded,
    //       but this doesn't seem to be exposed via the ts type definitions. 
    processEntities: false,
    tagValueProcessor: (tagName, a) => { 
        if (a) {
            return a.toString()
                .replace(/</g, '&lt;')
                .replace(/>/g, '&gt;');
        }
        return a; 
    },
});
```

**Comment by amitguptagwl** [comment]
Yes, it is somehow missed from typings.
